core/build.rs

```
@@ -1,8 +1,33 @@
+use std::path::{Path, PathBuf};
+
 fn main() {
-    lalrpop::Configuration::new()
-        .use_cargo_dir_conventions()
-        .process_file("src/parser/grammar.lalrpop")
-        .unwrap();
+    let checked_in_grammar_path = Path::new(&concat!(
+        env!("CARGO_MANIFEST_DIR"),
+        "/src/parser/grammar.rs"
+    ));
+
+    let out_dir = PathBuf::from(std::env::var_os("OUT_DIR").expect("missing OUT_DIR variable"));
+    let out_parser_dir = out_dir.join("parser");
+    std::fs::create_dir_all(&out_parser_dir).expect("failed to create $OUT_DIR/parser");
+    // Running lalrpop can be expensive. When building from git, we generate the parser
+    // in this build script, but when publishing the crate we add the generated
+    // parser to the published crate at `src/parser/grammar.rs`.
+    //
+    // In order to have this build script work for both the published crate and the git
+    // version, we try to copy `src/parser/grammar.rs` into the same location in $OUT_DIR
+    // that lalrpop would generate the grammar. If that copy fails because `src/parser/grammar.rs`
+    // doesn't exist, we're probably building from git and so we generate the grammar.
+    match std::fs::copy(checked_in_grammar_path, out_parser_dir.join("grammar.rs")) {
+        Ok(_) => {}
+        Err(e) if e.kind() == std::io::ErrorKind::NotFound => {
+            lalrpop::Configuration::new()
+                .use_cargo_dir_conventions()
+                .process_file("src/parser/grammar.lalrpop")
+                .unwrap();
+        }
+        Err(e) => panic!("{e}"),
+    }
+    println!("cargo:rerun-if-changed=src/parser/grammar.lalrpop");
```

If I understand correctly, we don't perform the sha3 check here, so we always assume the provided grammar file is up to date. It's not necessarily an issue, if we pack it up when packaging for crates.io, but I wonder if this extra check couldn't avoid some hard to track issues (maybe some other use cases around providing the generated parser will follow in the future, e.g. get a ready to work dev environment?)

---

> Are you commenting specifically on the "cargo:rerun-if-changed" line, or the general philosophy of copying the generated grammar to the output directory if it's present without checking hashes?

I just used the last line but it was about the general philosophy, yeah. Maybe I'm just overthinking it, but I just have an impression that having a branching like this in the build script that can silently ignores the `.lalrpop` file has a very low but non null probability of biting someone, some day. However, you're right that right now for our release use-case it also makes sense - and it's simpler - to take a pre-generated parser if it's there, unconditionally.

Maybe a very cheap way to prevent any issue would be simply to output something during build if we're using the checked file? Like "Found a pre-generated LALRPOP parser, using it and ignoring the grammar" or something like that, so that it's visible in the log?

---

Sure, I added a message in both branches.